graphql_compiler/query_formatting/graphql_formatting.py

```diff
@@ -1,7 +1,43 @@
 # Copyright 2017 Kensho Technologies, Inc.
-import re
+from graphql import parse
+from graphql.language.printer import PrintingVisitor, join, wrap
+from graphql.language.visitor import visit
 
-from graphql import parse, print_ast
+from ..schema import DIRECTIVES
+
+
+def pretty_print_graphql(query, use_four_spaces=True):
+    """Take a GraphQL query, pretty print it, and return it."""
+    output = visit(parse(query), CustomPrintingVisitor())
+    if use_four_spaces:
+        return fix_indentation_depth(output)
+    return output
+
+
+DIRECTIVES_BY_NAME = {d.name: d for d in DIRECTIVES}
+
+
+class CustomPrintingVisitor(PrintingVisitor):
+    # Directives are easier to read if their arguments appear in the order in
+    # which we defined them in the schema. For example, @filter directives are
+    # much easier to read if the operation comes before the values. The
+    # arguments of the directives specified in the schema are defined as
+    # OrderedDicts which allows us to sort the provided arguments to match.
+    def leave_Directive(self, node, *args):
+        """Call when exiting a directive node in the ast."""
+        args = node.arguments
+        arg_names = [a.split(':', 1)[0] for a in args]
+
+        directive = DIRECTIVES_BY_NAME.get(node.name)
+        if directive:
+            sorted_args = []
+            for defined_arg in directive.args.keys():
+                if defined_arg in arg_names:
+                    arg = args.pop(arg_names.index(defined_arg))
```

Comment:
`args` comes from `node.arguments` and is being mutated here. It's generally a bad idea to mutate a function's arguments. How about either making a copy, or simply using a `set` to remember which arguments were already processed?